File: 2015/day01.cpp
// https://adventofcode.com/2015/day/1

#include "../common/common.h"

#ifndef AOC_IMPORT_STD
#include <fstream>
#include <print>
#include <string>
#include <string_view>
#endif

std::string parse(const std::string& filename) {
  return aoc::read_single_line(filename);
}

template <bool check_position>
int solve_case(std::string_view directions) {
  int floor = 0;
  [[maybe_unused]] int position = 0;
  for (char c : directions) {
    floor += static_cast<int>(c == '(') - static_cast<int>(c == ')');
    if constexpr (check_position) {
      ++position;
      if (floor == -1) {
        break;
      }
    }
  }
  return check_position ? position : floor;
}

int main() {
  std::println("Part 1");
  const auto example = parse("day01.example");
  AOC_EXPECT_RESULT(-1, solve_case<false>(example));
  const auto input = parse("day01.input");
  AOC_EXPECT_RESULT(280, solve_case<false>(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(5, solve_case<true>(example));
  AOC_EXPECT_RESULT(1797, solve_case<true>(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2015/src/day01.rs
use std::fs::File;
use std::io::{self, BufRead, BufReader};

fn parse(filename: &str) -> io::Result<String> {
    let file = File::open(filename)?;
    let mut line = String::new();
    BufReader::new(file).read_line(&mut line)?;
    Ok(line)
}

fn solve_case<const CHECK_POSITION: bool>(directions: &str) -> i32 {
    let mut floor = 0;
    let mut position = 0;
    for c in directions.chars() {
        floor += (c == '(') as i32 - (c == ')') as i32;
        if CHECK_POSITION {
            position += 1;
            if floor == -1 {
                break;
            }
        }
    }
    if CHECK_POSITION { position } else { floor }
}

fn main() {
    println!("Part 1");
    let example = parse("day01.example").unwrap();
    aoc::expect_result!(-1, solve_case::<false>(&example));
    let input = parse("day01.input").unwrap();
    aoc::expect_result!(280, solve_case::<false>(&input));

    println!("Part 2");
    aoc::expect_result!(5, solve_case::<true>(&example));
    aoc::expect_result!(1797, solve_case::<true>(&input));
}
